group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java

```diff
@@ -400,6 +408,14 @@ public CoordinatorResult<OffsetDeleteResponseData, Record> deleteOffsets(
         return offsetMetadataManager.deleteOffsets(request);
     }
 
+    public CoordinatorResult<Void, Record> cleanupGroupMetadata() {
+        List<Record> records = new ArrayList<>();
+        offsetMetadataManager.cleanupOffsetMetadata(records, config.offsetsRetentionMs);
+        groupMetadataManager.cleanupGroupMetadata(records);
```

Comment:
It is a tad annoying that we have to iterate on all the groups with offsets in `cleanupOffsetMetadata` and on the groups in `cleanupGroupMetadata`. Would it be possible to iterate on all groups only once?

It is also annoying that we have to iterate on all the groups as well because it won't scale well. We would need to build indices to avoid this but I am not sure if it is worth it. Have you already considered this?